teuthology/dispatcher/__init__.py

```diff
@@ -0,0 +1,172 @@
+    # run dispatcher in job supervisor mode if --supervisor passed
+    if args["--supervisor"]:
+        return supervisor.main(args)
+
+    verbose = args["--verbose"]
+    tube = args["--tube"]
+    log_dir = args["--log-dir"]
+    archive_dir = args["--archive-dir"]
+
+    # setup logging for disoatcher in {log_dir}
+    loglevel = logging.INFO
+    if verbose:
+        loglevel = logging.DEBUG
+    log.setLevel(loglevel)
+    log_file_path = os.path.join(log_dir, f"dispatcher.{tube}.{os.getpid()}")
+    setup_log_file(log_file_path)
+    install_except_hook()
+
+    load_config(archive_dir=archive_dir)
+
+    connection = beanstalk.connect()
+    beanstalk.watch_tube(connection, tube)
+    result_proc = None
+
+    if teuth_config.teuthology_path is None:
+        fetch_teuthology('master')
+    fetch_qa_suite('master')
+
+    keep_running = True
+    while keep_running:
+        # Check to see if we have a teuthology-results process hanging around
+        # and if so, read its return code so that it can exit.
+        if result_proc is not None and result_proc.poll() is not None:
+            log.debug("teuthology-results exited with code: %s",
+                      result_proc.returncode)
+            result_proc = None
+
+        if sentinel(restart_file_path):
+            restart()
+        elif sentinel(stop_file_path):
+            stop()
+
+        load_config()
+
+        job = connection.reserve(timeout=60)
+        if job is None:
+            continue
+
+        # bury the job so it won't be re-run if it fails
+        job.bury()
+        job_id = job.jid
+        log.info('Reserved job %d', job_id)
+        log.info('Config is: %s', job.body)
+        job_config = yaml.safe_load(job.body)
+        job_config['job_id'] = str(job_id)
+
+        if job_config.get('stop_worker'):
+            keep_running = False
+
+        job_config, teuth_bin_path = prep_job(
+            job_config,
+            log_file_path,
+            archive_dir,
+        )
+
+        # lock machines but do not reimage them
+        if 'roles' in job_config:
+            job_config = lock_machines(job_config)
+
+        run_args = [
+            os.path.join(teuth_bin_path, 'teuthology-dispatcher'),
+            '--supervisor',
+            '-v',
+            '--bin-path', teuth_bin_path,
+            '--archive-dir', archive_dir,
+        ]
+
+        # Create run archive directory if not already created and
+        # job's archive directory
+        create_job_archive(job_config['name'],
+                           job_config['archive_path'],
+                           archive_dir)
+        job_config_path = os.path.join(job_config['archive_path'], 'orig.config.yaml')
+
+        # Write initial job config in job archive dir
+        with open(job_config_path, 'w') as f:
+            yaml.safe_dump(job_config, f, default_flow_style=False)
+
+        run_args.extend(["--job-config", job_config_path])
+        job_proc = subprocess.Popen(run_args)
```

Comment:
looks like we need some error handling here in case spawning the process fails for some reason, e.g.:

```Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_wip-minimal-dispatcher/virtualenv/bin/teuthology-dispatcher", line 33, in <module>
    sys.exit(load_entry_point('teuthology', 'console_scripts', 'teuthology-dispatcher')())
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_wip-minimal-dispatcher/scripts/dispatcher.py", line 33, in main
    teuthology.dispatcher.main(args)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_wip-minimal-dispatcher/teuthology/dispatcher/__init__.py", line 146, in main
    job_proc = subprocess.Popen(run_args)
  File "/usr/lib/python3.6/subprocess.py", line 729, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.6/subprocess.py", line 1364, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/bin/teuthology-dispatcher': '/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/bin/teuthology-dispatcher'```

To keep things simple we could log the error, delete the job, and continue. Ideally we could mark the job failed with the exception we got too.

Reply:
Ah right. Now logs error message along with the exception raised, unlocks targets and reports status.